endpointmanager/cmd/main.go

```diff
@@ -1,9 +1,88 @@
 package main
 
 import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/onc-healthit/lantern-back-end/endpointmanager/pkg/config"
+	"github.com/onc-healthit/lantern-back-end/endpointmanager/pkg/endpointmanager"
+	"github.com/onc-healthit/lantern-back-end/endpointmanager/pkg/endpointmanager/postgresql"
+	"github.com/onc-healthit/lantern-back-end/lanternmq"
+	"github.com/onc-healthit/lantern-back-end/lanternmq/pkg/accessqueue"
 	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
 )
 
+func failOnError(err error) {
+	if err != nil {
+		log.Fatalf("%s", err)
+	}
+}
+
+// getEndptsAndSend gets the current list of endpoints from the database and sends each one to the given queue
+// it continues to repeat this action every time the given interval period has passed
+func getEnptsAndSend(
+	ctx context.Context,
+	qName string,
+	qInterval int,
+	store endpointmanager.FHIREndpointStore,
+	mq *lanternmq.MessageQueue,
+	channelID *lanternmq.ChannelID) {
+
+	for {
+		listOfEndpoints, err := store.GetAllFHIREndpoints(ctx)
+		failOnError(err)
+
+		for i, endpt := range *listOfEndpoints {
+			if i%10 == 0 {
+				log.Infof("Processed %d/%d messages", i, len(*listOfEndpoints))
+			}
+
+			msgBytes, err := json.Marshal(endpt)
+			failOnError(err)
+			msgStr := string(msgBytes)
+			err = accessqueue.SendToQueue(ctx, msgStr, mq, channelID, qName)
+			failOnError(err)
+		}
+
+		log.Infof("Waiting %d minutes", qInterval)
+		time.Sleep(time.Duration(qInterval) * time.Minute)
+	}
+}
+
 func main() {
 	log.Info("Started the endpoint manager.")
+
+	err := config.SetupConfig()
+	failOnError(err)
+
+	store, err := postgresql.NewStore(viper.GetString("dbhost"), viper.GetInt("dbport"), viper.GetString("dbuser"), viper.GetString("dbpassword"), viper.GetString("dbname"), viper.GetString("dbsslmode"))
+	failOnError(err)
+	log.Info("Successfully connected to DB!")
+
+	// Set up the queue for sending messages to capabilityquerier and networkstatsquerier
+	capQName := viper.GetString("enptinfo_capquery_qname")
+	mq, channelID, err := accessqueue.ConnectToServerAndQueue(viper.GetString("quser"), viper.GetString("qpassword"), viper.GetString("qhost"), viper.GetString("qport"), capQName)
+	failOnError(err)
+	log.Info("Successfully connected to capabilityquerier Queue!")
+
+	netQName := viper.GetString("enptinfo_netstats_qname")
+	mq, channelID, err = accessqueue.ConnectToQueue(mq, channelID, netQName)
+	failOnError(err)
+	log.Info("Successfully connected to networkstatsquerier Queue!")
+
+	// Infinite query loop
+	var wg sync.WaitGroup
+	ctx := context.Background()
+	wg.Add(1)
+	capInterval := viper.GetInt("capquery_qryintvl")
+	go getEnptsAndSend(ctx, capQName, capInterval, store, &mq, &channelID)
+
+	wg.Add(1)
+	netInterval := viper.GetInt("endptqry_query_interval")
+	go getEnptsAndSend(ctx, netQName, netInterval, store, &mq, &channelID)
+
+	wg.Wait()
```

Comment:
it doesn't look like the wait group is being passed in anywhere to get signaled that it's done.
it needs something to signal that what you added to it is Done so that `wg.Wait()` will end at some point. checkout how it's used in `capabilityquerier/pkg/capabilityquerier/queueworkers.go`

Reply:
i was using this as a way to prevent the main function from exiting that i saw recommended so both go routines could run forever on their respective intervals. I'm not sure at what point it would end, since it seems like you'd probably want to keep running the routine in all of the potential error cases in `getEndptsAndSend`. Is there another way you know of for keeping a goroutine running forever that would be more preferable? 

Now that I'm thinking about it though, all of the failOnErrors in that function should probably either just be logs or pushed to an error channel that is printing them out in the main function... What do you think?

Reply:
yeah agree w error channel, which will also keep the main function open forever. See `ReceiveCapabilityStatements` in `pkg/capabilityreceiver.go`:

```golang
// ReceiveCapabilityStatements connects to the given message queue channel and receives the capability
// statements from it. It then adds the capability statements to the given store.
func ReceiveCapabilityStatements(ctx context.Context,
	store *postgresql.Store,
	messageQueue lanternmq.MessageQueue,
	channelID lanternmq.ChannelID,
	qName string) error {

	args := make(map[string]interface{})
	args["store"] = store
	args["ctx"] = ctx

	messages, err := messageQueue.ConsumeFromQueue(channelID, qName)
	if err != nil {
		return err
	}

	errs := make(chan error)
	go messageQueue.ProcessMessages(messages, saveMsgInDB, &args, errs)

	for elem := range errs {
		log.Warn(elem)
	}

	return nil
}
```

if you were to use the work group, i think you'd still want to pass it in and have a `defer wg.Done()` call in your go routine so that if it does get some sort of fatal error, that gets signaled somehow. that may be what we'll want to do long term when we set all of these up as individual go routines.

Reply:
that's fair, i'll add that then!

Reply:
👍